collagraph/cgx/cgx.py

```diff
@@ -54,218 +54,339 @@ class Foo(cg.Component):
+                def _lookup(self, name, cache={}):
+                    # Note that the default value of cache is using the fact
+                    # that defaults are created at function definition, so
+                    # the cache is actually a 'global' object that is shared
+                    # between method calls and thus is suited to serve as
+                    # a cache for storing the method used for looking up the
+                    # value.
+                    if method := cache.get(name):
+                        return method(self, name)
+
+                    def props_lookup(self, name):
+                        return self.props[name]
+
+                    def state_lookup(self, name):
+                        return self.state[name]
+
+                    def self_lookup(self, name):
+                        return getattr(self, name)
+
+                    def global_lookup(self, name):
+                        return globals()[name]
+
+                    if name in self.props:
+                        cache[name] = props_lookup
+                    elif name in self.state:
+                        cache[name] = state_lookup
+                    elif hasattr(self, name):
+                        cache[name] = self_lookup
+                    elif name in globals():
+                        cache[name] = global_lookup
+                    else:
+                        raise NameError(f"name '{name}' is not defined")
+                    return _lookup(self, name)
+                """
+            ),
+            mode="exec",
+        ).body[0],
+    )
 
-    # Construct a render function from the template block
-    template_root_node = parser.root.child_with_tag("template").children[0]
-    compiled_tree = template_root_node.compile(preliminary_component_type, context)
+    # Find the first ClassDef and assume that it is the
+    # component that is defined in the SFC
+    component_def = None
+    for node in script_tree.body:
+        if isinstance(node, ast.ClassDef):
+            component_def = node
+            break
+
+    # Create render function as AST and inject into the ClassDef
+    render_tree = create_ast_render_function(
+        parser.root.child_with_tag("template").children[0]
+    )
+    component_def.body.append(render_tree)
+
+    # Because we modified the AST significantly we need to call an AST
+    # method to fix any `lineno` and `col_offset` attributes of the nodes
+    ast.fix_missing_locations(script_tree)
+
+    # Compile the tree into a code object (module)
+    code = compile(script_tree, filename="<ast>", mode="exec")
+    # Execute the code as module and pass a dictionary that will capture
+    # the global and local scope of the module
+    module_namespace = {}
+    exec(code, module_namespace)
+
+    # Check that the class definition is an actual subclass of Component
+    component_class = module_namespace[component_def.name]
+    if not issubclass(component_class, Component):
```

Comment:
seems like it's pretty easy to trigger this if you define e.g. a dataclass or something else in your cgx file, before your component class definition

Reply:
Yes. Without evaluating/executing the SFC module, it's not possible to know whether a ClassDef is a sub(sub)class of Component, so I think we should just document that the first class definition in your SFC should be a component sub(sub)class.

Reply:
Or... The last? I think that might be more practical, if you think about reasons to define more classes in a cgx file

Reply:
👍makes sense

Reply:
I've also added a test to make this behaviour a bit more 'documented'.